File: src/main/java/gg/mineral/practice/commands/config/ArenaCommand.java
package gg.mineral.practice.commands.config;

import java.util.Locale;

import org.bukkit.craftbukkit.v1_8_R3.entity.CraftPlayer;
import org.bukkit.entity.Player;

import gg.mineral.practice.arena.Arena;
import gg.mineral.practice.commands.PlayerCommand;
import gg.mineral.practice.managers.ArenaManager;
import gg.mineral.practice.util.PlayerUtil;
import gg.mineral.practice.util.messages.CC;
import gg.mineral.practice.util.messages.impl.ChatMessages;
import gg.mineral.practice.util.messages.impl.ErrorMessages;
import gg.mineral.practice.util.messages.impl.UsageMessages;

import lombok.val;

public class ArenaCommand extends PlayerCommand {

    public ArenaCommand() {
        super("arena", "practice.config");
    }

    private static final String CREATE = "create", SPAWN = "spawn", SETDISPLAY = "setdisplay", LIST = "list",
            TELEPORT = "teleport", TP = "tp", REMOVE = "remove", DELETE = "delete", PLACEHOLDER = "%arena%";

    private static final int CREATE_ARGS = 2, SPAWN_ARGS = 3, DISPLAY_ARGS = 2, TELEPORT_ARGS = 2, DELETE_ARGS = 2;

    @Override
    public void execute(final Player player, final String[] args) {
        switch (args.length > 0 ? args[0].toLowerCase(Locale.ROOT) : "") {
            default -> {
                ChatMessages.ARENA_COMMANDS.send(player);
                ChatMessages.ARENA_CREATE.send(player);
                ChatMessages.ARENA_SPAWN.send(player);
                ChatMessages.ARENA_DISPLAY.send(player);
                ChatMessages.ARENA_LIST.send(player);
                ChatMessages.ARENA_TP.send(player);
                ChatMessages.ARENA_DELETE.send(player);
            }
            case CREATE -> {
                if (args.length < CREATE_ARGS) {
                    UsageMessages.ARENA_CREATE.send(player);
                    return;
                }

                val arenaName = args[1];

                if (ArenaManager.getArenaByName(arenaName) != null) {
                    ErrorMessages.ARENA_ALREADY_EXISTS.send(player);
                    return;
                }

                val arena = new Arena(arenaName, ArenaManager.CURRENT_ID++);
                arena.setDefaults();
                ArenaManager.registerArena(
                        arena);
                ChatMessages.ARENA_CREATED.clone().replace(PLACEHOLDER, arenaName).send(player);
            }
            case SPAWN -> {
                if (args.length < SPAWN_ARGS) {
                    UsageMessages.ARENA_SPAWN.send(player);
                    return;
                }

                val arenaName = args[1];
                val arena = ArenaManager.getArenaByName(arenaName);

                if (arena == null) {
                    ErrorMessages.ARENA_DOES_NOT_EXIST.send(player);
                    return;
                }

                switch (args[2].toLowerCase(Locale.ROOT)) {
                    case "1" ->
                        arena.setLocation1(player.getLocation());
                    case "2" -> arena.setLocation2(player.getLocation());
                    case "waiting" -> arena.setWaitingLocation(player.getLocation());
                    default -> {
                        UsageMessages.ARENA_SPAWN.send(player);
                        return;
                    }
                }

                ChatMessages.ARENA_SPAWN_SET.clone().replace(PLACEHOLDER, arenaName).send(player);
            }
            case SETDISPLAY -> {
                if (args.length < DISPLAY_ARGS) {
                    UsageMessages.ARENA_DISPLAY.send(player);
                    return;
                }

                val arenaName = args[1];
                val arena = ArenaManager.getArenaByName(arenaName);

                if (arena == null) {
                    ErrorMessages.ARENA_DOES_NOT_EXIST.send(player);
                    return;
                }

                arena.setDisplayItem(player.getItemInHand());

                if (args.length > DISPLAY_ARGS)
                    arena.setDisplayName(args[2].replace("&", "§"));

                ChatMessages.ARENA_DISPLAY_SET.clone().replace(PLACEHOLDER, arenaName).send(player);
            }
            case LIST -> {
                val sb = new StringBuilder(CC.GRAY + "[");

                val iterator = ArenaManager.getArenas().values().iterator();

                while (iterator.hasNext()) {
                    val a = iterator.next();
                    sb.append(CC.GREEN + a.getName());

                    if (iterator.hasNext())
                        sb.append(CC.GRAY + ", ");
                }

                sb.append(CC.GRAY + "]");

                player.sendMessage(sb.toString());
            }
            case TELEPORT, TP -> {
                if (args.length < TELEPORT_ARGS) {
                    UsageMessages.ARENA_TP.send(player);
                    return;
                }

                val arena = ArenaManager.getArenaByName(args[1]);

                if (arena == null) {
                    ErrorMessages.ARENA_DOES_NOT_EXIST.send(player);
                    return;
                }

                try {
                    PlayerUtil.teleport((CraftPlayer) player, arena.getLocation1());
                } catch (Exception e) {
                    ErrorMessages.CANNOT_TELEPORT_TO_ARENA.send(player);
                    e.printStackTrace();
                }

            }
            case DELETE, REMOVE -> {
                if (args.length < DELETE_ARGS) {
                    UsageMessages.ARENA_DELETE.send(player);
                    return;
                }

                val arena = ArenaManager.getArenaByName(args[1]);

                if (arena == null) {
                    ErrorMessages.ARENA_DOES_NOT_EXIST.send(player);
                    return;
                }

                ArenaManager.remove(arena);
                val arenaName = args[1];
                ChatMessages.ARENA_DELETED.clone()
                        .replace(PLACEHOLDER, arenaName).send(player);
            }
        }
    }
}

File: src/main/kotlin/gg/mineral/practice/commands/config/ArenaCommand.kt
package gg.mineral.practice.commands.config

import dev.rollczi.litecommands.annotations.argument.Arg
import dev.rollczi.litecommands.annotations.command.Command
import dev.rollczi.litecommands.annotations.context.Context
import dev.rollczi.litecommands.annotations.execute.Execute
import dev.rollczi.litecommands.annotations.permission.Permission
import gg.mineral.practice.arena.Arena
import gg.mineral.practice.arena.ArenaImpl
import gg.mineral.practice.arena.EventArena
import gg.mineral.practice.entity.appender.CommandSenderAppender
import gg.mineral.practice.managers.ArenaManager
import gg.mineral.practice.managers.ArenaManager.arenas
import gg.mineral.practice.managers.ArenaManager.getArenaByName
import gg.mineral.practice.managers.ArenaManager.registerArena
import gg.mineral.practice.managers.ArenaManager.remove
import gg.mineral.practice.util.PlayerUtil
import gg.mineral.practice.util.messages.CC
import gg.mineral.practice.util.messages.impl.ChatMessages
import gg.mineral.practice.util.messages.impl.ErrorMessages
import gg.mineral.practice.util.messages.impl.UsageMessages
import gg.mineral.practice.util.world.SpawnLocation
import gg.mineral.practice.util.world.appender.LocationAppender
import org.bukkit.command.CommandSender
import org.bukkit.craftbukkit.v1_8_R3.entity.CraftPlayer
import org.bukkit.entity.Player
import java.util.*

@Command(name = "arena")
@Permission("practice.config")
class ArenaCommand : CommandSenderAppender, LocationAppender {
    @Execute(name = "create")
    fun executeCreate(@Context sender: CommandSender, @Arg name: String) {
        if (getArenaByName(name) != null) {
            sender.send(ErrorMessages.ARENA_ALREADY_EXISTS)
            return
        }

        registerArena(ArenaImpl(name, ArenaManager.CURRENT_ID++))
        sender.send(ChatMessages.ARENA_CREATED.clone().replace(PLACEHOLDER, name))
    }

    @Execute(name = "spawn")
    fun executeSpawn(
        @Context sender: CommandSender,
        @Context location: SpawnLocation,
        @Arg arena: Arena,
        @Arg spawn: String
    ) {
        when (spawn.lowercase()) {
            "1" -> arena.location1 = location
            "2" -> arena.location2 = location
            "waiting" -> (arena as? EventArena)?.waitingLocation = location
            else -> {
                sender.send(UsageMessages.ARENA_SPAWN)
                return
            }
        }

        sender.send(ChatMessages.ARENA_SPAWN_SET.clone().replace(PLACEHOLDER, arena.name))
    }

    @Execute(name = "edit")
    fun executeEdit() {

    }

    @Execute(name = "setdisplay", aliases = ["display"])
    fun executeSetDisplay(@Context player: Player, @Arg arena: Arena, @Arg displayName: Optional<String>) {
        arena.displayItem = player.itemInHand
        displayName.ifPresent { arena.displayName = it.replace("&", "§") }
        player.send(ChatMessages.ARENA_DISPLAY_SET.clone().replace(PLACEHOLDER, arena.name))
    }

    @Execute(name = "list")
    fun executeList(@Context sender: CommandSender) {
        val sb = StringBuilder(CC.GRAY + "[")

        val iterator = arenas.values.iterator()

        while (iterator.hasNext()) {
            iterator.next()?.let {
                sb.append(CC.GREEN).append(it.name)
                if (iterator.hasNext()) sb.append(CC.GRAY).append(", ")
            }
        }

        sb.append(CC.GRAY).append("]")

        sender.sendMessage(sb.toString())
    }

    @Execute(name = "teleport", aliases = ["tp"])
    fun executeTeleport(@Context player: Player, @Arg arena: Arena) {
        try {
            val world = arena.generateBaseWorld()
            PlayerUtil.teleport(player as CraftPlayer, arena.location1.bukkit(world))
        } catch (e: Exception) {
            ErrorMessages.CANNOT_TELEPORT_TO_ARENA.send(player)
            e.printStackTrace()
        }
    }

    @Execute(name = "remove", aliases = ["delete"])
    fun executeRemove(@Context sender: CommandSender, @Arg arena: Arena) {
        remove(arena)
        sender.send(ChatMessages.ARENA_DELETED.clone().replace(PLACEHOLDER, arena.name))
    }

    @Execute(aliases = ["help"])
    fun executeHelp(@Context sender: CommandSender) {
        sender.send(
            ChatMessages.ARENA_COMMANDS,
            ChatMessages.ARENA_CREATE,
            ChatMessages.ARENA_SPAWN,
            ChatMessages.ARENA_DISPLAY,
            ChatMessages.ARENA_LIST,
            ChatMessages.ARENA_TP,
            ChatMessages.ARENA_DELETE
        )
    }

    companion object {
        private const val PLACEHOLDER = "%arena%"
    }
}
